controllers/spiaccesstokenbinding_controller.go

```diff
@@ -564,27 +677,141 @@ func (r *SPIAccessTokenBindingReconciler) syncSecret(ctx context.Context, sp ser
 	_, obj, err := r.syncer.Sync(ctx, nil, secret, secretDiffOpts)
 	if err != nil {
 		r.updateBindingStatusError(ctx, binding, api.SPIAccessTokenBindingErrorReasonTokenSync, err)
-		return api.TargetObjectRef{}, fmt.Errorf("failed to sync the secret with the token data: %w", err)
+		return nil, fmt.Errorf("failed to sync the secret with the token data: %w", err)
 	}
-	return toObjectRef(obj), nil
+	return obj.(*corev1.Secret), nil
+}
+
+func (r *SPIAccessTokenBindingReconciler) getOrCreateServiceAccount(ctx context.Context, binding *api.SPIAccessTokenBinding) (*corev1.ServiceAccount, error) {
+	spec := binding.Spec.ServiceAccount
+
+	if spec == nil {
+		return nil, nil
+	}
+
+	name := binding.Status.ServiceAccountName
+	if name == "" {
+		name = binding.Spec.ServiceAccount.Name
+	}
+
+	requestedLabels := spec.Labels
+	if requestedLabels == nil {
+		requestedLabels = map[string]string{}
+	}
+	requestedLabels[SPIAccessTokenBindingLinkLabel] = binding.Name
+
+	var err error
+	sa := &corev1.ServiceAccount{}
+	needsUpdate := false
+
+	if err = r.Client.Get(ctx, client.ObjectKey{Name: name, Namespace: binding.Namespace}, sa); err != nil {
+		if errors.IsNotFound(err) {
+			sa = &corev1.ServiceAccount{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:         name,
+					Namespace:    binding.Namespace,
+					GenerateName: binding.Spec.ServiceAccount.GenerateName,
+					Annotations:  binding.Spec.ServiceAccount.Annotations,
+					Labels:       requestedLabels,
+				},
+			}
+			err = r.Client.Create(ctx, sa)
+		}
+	}
+
+	// make sure all our labels have the values we want
+	if err == nil {
+		if sa.Labels == nil {
+			sa.Labels = map[string]string{}
+		}
+
+		for k, rv := range requestedLabels {
+			v, ok := sa.Labels[k]
+			if !ok || (ok && v != rv) {
+				needsUpdate = true
+			}
+			sa.Labels[k] = v
+		}
+
+		if needsUpdate {
+			err = r.Client.Update(ctx, sa)
+		}
+	}
+
+	if err != nil {
+		r.updateBindingStatusError(ctx, binding, api.SPIAccessTokenBindingErrorReasonTokenSync, err)
+		return nil, fmt.Errorf("failed to sync the configured service account of the binding: %w", err)
+	}
+
+	return sa, nil
+}
+
+func (r *SPIAccessTokenBindingReconciler) linkSecretToServiceAccount(ctx context.Context, secret *corev1.Secret, sa *corev1.ServiceAccount) error {
```

Comment:
doc and test please

Reply:
This has 100% code coverage in the integration tests.